components/pkg-export-helm/src/chart.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use clap;
+use std::fs;
+use std::io;
+use std::io::Write;
+
+use common::ui::{UI, Status};
+use export_docker;
+use export_docker::Result;
+use export_k8s::{Error, Manifest, ManifestJson};
+
+use chartfile::ChartFile;
+use values::Values;
+
+pub struct Chart<'a> {
+    name: String,
+    habitat_name: String,
+    chartfile: ChartFile,
+    manifest_template: ManifestJson,
+    values: Values,
+    ui: &'a mut UI,
+}
+
+impl<'a> Chart<'a> {
+    pub fn new_for_cli_matches(ui: &'a mut UI, matches: &clap::ArgMatches) -> Result<Self> {
+        if !matches.is_present("NO_DOCKER_IMAGE") {
+            export_docker::export_for_cli_matches(ui, &matches)?;
+        }
+        let manifest = Manifest::new_from_cli_matches(ui, &matches)?;
+        let name = matches
+            .value_of("CHART")
+            .unwrap_or(&manifest.habitat_name)
+            .to_string();
+        let version = matches.value_of("VERSION");
+        let description = matches.value_of("DESCRIPTION");
+        let chartfile = ChartFile::new(&name, version, description)?;
+
+        Chart::new_for_manifest(manifest, name, chartfile, ui)
+    }
+
+    fn new_for_manifest(
+        manifest: Manifest,
+        name: String,
+        chartfile: ChartFile,
+        ui: &'a mut UI,
+    ) -> Result<Self> {
+        let main = json!({
+            "metadata_name": "{{.Values.metadataName}}",
+            "habitat_name": "{{.Values.habitatName}}",
+            "image": "{{.Values.imageName}}",
+            "count": "{{.Values.instanceCount}}",
+            "service_topology": "{{.Values.serviceTopology}}",
+            "service_group": manifest.service_group.clone().map(|_| "{{.Values.serviceGroup}}"),
+            "config_secret_name": manifest.config_secret_name
+                .clone()
+                .map(|_| "{{.Values.configSecretName}}"),
+                "ring_secret_name": manifest.ring_secret_name
```

Comment:
Indentation is off here.

Reply:
not according rustfmt we are supposed to use with habitat (0.9.0). I'm hoping they'll at least switch to latest stable.

Reply:
Probably rustfmt gets confused. Probably it could do the right thing if you used:

```
"config_secret_name": if manifest.is_some() {
    Some("…")
} else {
    None
}
```

That would even save a `clone`, but that's a bit verbose. Maybe:

```
"config_secret_name": manifest.config_secret_name
    .clone()
    .and(Some("…"))
```

Reply:
i don't like either of those. I changed to:

```            "config_secret_name": manifest.config_secret_name
                .clone()
                .map(|_| "{{.Values.configSecretName}}"),
            "ring_secret_name": manifest.ring_secret_name
                .clone()
                .map(|_| "{{.Values.ringSecretName}}"),
            "bind": manifest.binds,
```

and rustfmt was ok with it.